prototype/large_scale_recommender/large_scale_recommender.ino

```diff
@@ -0,0 +1,118 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+
+#include "lu_decomposition.h"
+#include "finite_difference.h"
+#include "gradient_descent.h"
+#include "matrix.h"
+#include "arduino_serial_port.h"
+
+#define LATENT_SIZE 2
+uint8_t feedbackSize = 0;
+double *feedbackMatrix;
+
+double* gradientMatrix;
+
+void extractUserMatrix(const double* lMatrix, double* outputMatrix)
+{
+  for(uint8_t i = 0; i < feedbackSize; ++i)
+  {
+    memcpy(&outputMatrix[i * LATENT_SIZE], &lMatrix[i * feedbackSize], LATENT_SIZE * sizeof(double));
+  }
+}
+
+void extractProductMatrix(const double* uMatrix, double* outputMatrix)
+{
+  memcpy(outputMatrix, uMatrix, LATENT_SIZE * feedbackSize * sizeof(double));
+}
+
+static gradient_real func(gradient_real* parameters)
+{
+  double sum = 0;
+  const int OFFSET = feedbackSize * LATENT_SIZE;
+  const double weight = 0.01;
+  for(uint8_t i = 0; i < feedbackSize; ++i)
+  {
+    for(uint8_t j = 0; j < feedbackSize; ++j)
+    {
+      double currentSum = 0;
+      for(uint8_t k = 0; k < LATENT_SIZE; ++k)
+      {
+        currentSum += parameters[coordToIndex(i, k, LATENT_SIZE)] * parameters[coordToIndex(j, k, LATENT_SIZE) + OFFSET];
+      }
+
+      if(feedbackMatrix[coordToIndex(i, j, feedbackSize)] != -1)
+      {
+        const double value = feedbackMatrix[coordToIndex(i, j, feedbackSize)] - currentSum;
+        sum += value * value;
+      }
+      else
+      {
+        sum += weight * currentSum * currentSum;
+      }
+    }
+  }
+  return fabs(sum);
+}
+
+static void dfunc(gradient_real* parameters, gradient_real* gradient)
+{
+  gradientApproximation(func, parameters, gradient, feedbackSize * LATENT_SIZE * 2, 0.0001, Central);
```

Comment:
Maybe explain what the 2 and the 0.0001 are in a comment or define as constants?